src/node_buffer.cc

```diff
@@ -1142,6 +1142,143 @@ void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+  /* emtpy buffer separator? Then skip indexing and fancy mapping! */
+  if (sep_buf_len == 0)
+    goto map_bytes;
+
+  for (size_t i = 0; i < js_buf_length; ++i) {
+    unsigned char byte = *((unsigned char *)&buf + i);
+    // check first occurence...
+    // ...and if separator > 1 check consecutive bytes for equalness.
+    if (sep_buf_len == 1 && byte == first_match_byte) {
+      index_arr.push_back(i);
+    } else if (byte == first_match_byte) {
+      for (size_t j = 0; j < sep_buf_len; ++j) {
+        unsigned char sep_byte = *((unsigned char *)&buf + i + j);
+        // only push to array when the whole separator got compared
+        if (sep_byte == sep_buf[j] && j == sep_buf_len - 1) {
+          index_arr.push_back(i);
+        }
+      }
+    }
+  }
+
+  res_buffers = v8::Array::New(env->isolate(), index_arr.size());
+  len = (size_t)index_arr.size();
+
+  for (size_t i = 0; i < len + 1; i++) {
+    size_t index = index_arr[i] - cursor;
+    size_t read_len = index;
+    // handle last iteration, REVIEW(eljefedelrodeodeljefe): simplify
+    if (len == i)
+      read_len = js_buf_length - cursor;
+
+    char* data = (char*)malloc(read_len * sizeof(char));
+    memcpy(data, buf + cursor, read_len);
+
+    obj = Buffer::New(env->isolate(), data, read_len).ToLocalChecked();
+    res_buffers->Set(i, obj);
+    cursor = index_arr[i] + sep_buf_len;
+  }
+  goto return_array;
+
+map_bytes:
+  res_buffers = v8::Array::New(env->isolate(), js_buf_length);
+
+  for (size_t i = 0; i < js_buf_length; i++) {
+    char* data = (char*)malloc(1 * sizeof(char));
+    memcpy(data, buf + i, 1);
+
+    obj = Buffer::New(env->isolate(), data, 1).ToLocalChecked();
+    res_buffers->Set(i, obj);
+  }
+
+return_array:
+  persistent.Reset();
+  return args.GetReturnValue().Set(res_buffers);
+}
+
+void Join(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  Local<v8::Array> arr = args[0].As<v8::Array>();
+  char* data;
+  bool has_insert = !args[2]->IsNull();
+  // local fixed copies, since spreading insert is optional
+  size_t ins_length;
+  char* ins_data;
+  // count resulting buffer size based on every buffer in data length in array
+  size_t num_buffers = (size_t)args[1]->IntegerValue();
+  size_t buffers_size = 0;
+  for (size_t i = 0; i < num_buffers; i++) {
+    THROW_AND_RETURN_UNLESS_BUFFER(env, arr->Get(i));
+    buffers_size += arr->Get(i).As<Uint8Array>()->Length();
+  }
+  // handle insert here. Increase length of resulting buffer
+  // minus last insert to match Array.prototype.join behavior
+  if (has_insert) {
+    SPREAD_ARG(args[2], insert);
+    buffers_size += num_buffers * insert_length - insert_length; // no last ins
+    ins_length = insert_length;
+    // allocate and copy the insert for use in second loop
+    ins_data = (char*)malloc(buffers_size * sizeof(char));
+    memcpy(ins_data, insert_data, insert_length);
+  }
+
+  data = (char*)malloc(buffers_size * sizeof(char));
```

Comment:
* `sizeof(char)` is defined to be 1 – you can still do this if you think it helps readability, but I’m not sure it’s worth it
* `static_cast<char*>(…)` is preferred to C-style casts in the codebase
* It might be helpful to check that `malloc()` does not return `nullptr` (`CHECK` should suffice, but without it, users might just see `Segmentation fault` when it fails).